File: components/notifications-service/server/test/service_test.exs
defmodule ServiceTest do
  use ExUnit.Case
  import Mock

  alias Notifications.{Service, Dispatcher, CCRFailure, Event, Rule, Data, 
                      URLValidationRequest, URLValidationResponse, UsernamePassword, 
                      SecretId, Empty}
  alias Notifications.{RuleListResponse, RuleAddResponse, RuleGetResponse,
                      RuleUpdateResponse, RuleDeleteResponse, RuleIdentifier, 
                      WebhookSender, Data.SecretStore}

  describe "notify" do
    test "Unpacks and sends the notification over to Dispatcher" do
      with_mocks([
        {Dispatcher, [], [process_notification: fn(_payload, _id) -> :ok  end]}
      ]) do
        ccr_failure = CCRFailure.new(run_id: "a")
        event = %Event{event: {:CCRFailure, ccr_failure}, id: "someid"}
        %type{} = Service.notify(event, :ignore)
        assert type == Notifications.Response
        assert called Dispatcher.process_notification(ccr_failure, "id:someid")
      end
    end
  end

  describe "version" do
    # TODO Pended because the application behaves differently in test than it does
    # when running - so the version lookup in the test env fails, while on a live instance
    # it does the right thing.
    @tag :pend
    test "it returns the correct release version" do
      version = String.trim(File.read!("../VERSION"))
      response = Service.version(VersionRequest.new, :ignore)
      assert version == response.version
    end
  end

  describe "list_rules" do
    test "that when rules exists they are returned" do
      with_mocks([
        {Data.Store, [], [get_rules: fn() -> {:ok, [rule1(), rule2_slack()]} end]},
      ]) do
        response = Service.list_rules(:ignore, :ignore)
        expected = RuleListResponse.new(code: 0, messages: [],
                                        rules: [rule1(), rule2_slack()])

        assert response ==  expected
      end
    end

    test "that when no rules exist, the response includes an empty list of rules" do
      with_mocks([
        {Data.Store, [], [get_rules: fn() -> {:ok, []} end]},
      ]) do
        response = Service.list_rules(:ignore, :ignore)
        assert response == RuleListResponse.new(code: 0, messages: [], rules: [])
      end
    end
  end

  describe "get_rule" do
    test "replies with not found error when the requested rule is not found" do
      with_mocks([
        {Data.Store, [], [get_rule: fn(_) -> {:ok, []} end]},
      ]) do
        rule_request = %RuleIdentifier{id: "rule1"}
        response = Service.get_rule(rule_request, :ignore)
        assert response.code == RuleGetResponse.Code.value(:NOT_FOUND)
        assert length(response.messages) == 1
      end
    end

    test "replies with the rule when the rule exists" do
      with_mocks([
        {Data.Store, [], [get_rule: fn(%{id: "rule2"}) -> {:ok, [rule2_slack()]} end ]},
      ]) do
        rule_request = %RuleIdentifier{id: "rule2"}
        response = Service.get_rule(rule_request,  :ignore)
        expected = %RuleGetResponse{code: 0, messages: [],
                                    rule: rule2_slack()}
        assert response == expected
      end
    end
  end

  describe "delete_rule" do
    test "replies with not found error if the rule to be deleted doesn't exist" do
      with_mocks([
        {Data.Store, [], [delete_rule: fn(%{id: "rule1"}) -> :not_found end ]}
      ]) do
        rule_request = %RuleIdentifier{id: "rule1"}
        response = Service.delete_rule(rule_request, :ignore)
        assert response.code == RuleAddResponse.Code.value(:NOT_FOUND)
        assert length(response.messages) == 1
      end
    end

    test "deletes and replies with successful response when an existing rule is deleted" do
      with_mocks([
        {Data.Store, [], [delete_rule: fn(%{id: "rule1_ID"}) -> :ok end]},
      ]) do
        rule_request = %RuleIdentifier{id: "rule1_ID"}
        response = Service.delete_rule(rule_request,  :ignore)
        expected = %RuleDeleteResponse{code: 0, messages: []}
        assert response == expected
      end
    end
  end

  describe "add_rule" do
    test "that it saves a valid rule and replies correctly" do
      with_mocks([
        {Data.Store, [], [add_rule: fn(_) -> {:ok, "abc123"} end]},
      ]) do
        rule_request = %{rule1() | id: ""}
        response = Service.add_rule(rule_request,  :ignore)
        assert response.code == 0
        assert response.messages == []
        assert response.id == "abc123"
      end
    end

    test "that it rejects the rule with duplicate name if a rule with the proposed name already exists" do
      with_mocks([
        {Data.Store, [], [add_rule: fn(_) -> {:error, :conflict} end]},
      ]) do
        rule_request = %{rule1() | id: ""}
        response = Service.add_rule(rule_request,  :ignore)
        assert response.code == RuleAddResponse.Code.value(:DUPLICATE_NAME)
        assert length(response.messages) == 1
      end
    end


    test "that it rejects the rule with validation errors if the rule is invalid" do
      response = Service.add_rule(Rule.new(),  :ignore)
      assert response.code == RuleAddResponse.Code.value(:VALIDATION_ERROR)
      # The specific validations are covered in Validator.Rule.Test
      assert length(response.messages) > 0
    end
  end

  describe "update_rule" do
    test "replies correctly when a valid update is processed" do
      with_mocks([
        {Data.Store, [], [update_rule: fn(%Rule{id: "rule1_ID"}) -> :ok end]}
      ]) do
        response = Service.update_rule(rule1(),  :ignore)
        expected = RuleUpdateResponse.new()
        assert response == expected
      end
    end

    test "that it rejects the rule with duplicate name if a rule with that name exists" do
      update_rule_mock = fn(%Rule{id: "rule3_webhook_ID", name: "duplicate-name"}) ->
                           {:error, :conflict}
                         end
      with_mocks([
        {Data.Store, [], [update_rule: update_rule_mock]},
      ]) do
        update_request = %{rule3_webhook() | name: "duplicate-name"}
        response = Service.update_rule(update_request,  :ignore)
        assert response.code == RuleUpdateResponse.Code.value(:DUPLICATE_NAME)
        assert length(response.messages) == 1
      end
    end

    test "that it rejects the rule with validation errors if the rule is invalid" do
      update_request = Rule.new()
      response = Service.update_rule(update_request,  :ignore)
      assert response.code == RuleUpdateResponse.Code.value(:VALIDATION_ERROR)
      assert length(response.messages) > 0
    end
  end

  describe "validate_webhook" do
    test "with UsernamePassword case" do
      with_mocks([
        {WebhookSender, [], [validate_webhook: fn(_url, _username, _password) -> {:ok, {200, ""}} end]}
      ]) do
        username = "bob"
        url = "url"
        password = "super_secret"
        request = %URLValidationRequest{url: url, credentials: {:username_password, 
          %UsernamePassword{username: username, password: password}}}
        response = Service.validate_webhook(request, :ignore)
        expected = %URLValidationResponse{code: 0, messages: []}
        assert response == expected
        assert called WebhookSender.validate_webhook(url, username, password)
      end
    end

    test "with SecretId case" do
      username = "bob"
      url = "url"
      password = "super_secret"
      secret_id = "super_secret_id"
      with_mocks([
        {WebhookSender, [], [validate_webhook: fn(_url, _username, _password) -> {:ok, {200, ""}} end]},
        {SecretStore, [], [get_target_username_password: fn(_secret_id) -> {:ok, {username, password}} end]}
      ]) do
        request = %URLValidationRequest{url: url, credentials: {:secret_id, %SecretId{id: secret_id}}}
        response = Service.validate_webhook(request, :ignore)
        expected = %URLValidationResponse{code: 0, messages: []}
        assert response == expected
        assert called SecretStore.get_target_username_password(secret_id)
        assert called WebhookSender.validate_webhook(url, username, password)
      end
    end

    test "with Empty case" do
      with_mocks([
        {WebhookSender, [], [validate_webhook: fn(_url, _username, _password) -> {:ok, {200, ""}} end]}
      ]) do
        url = "url"
        request = %URLValidationRequest{url: url, credentials: {:none, %Empty{}}}
        response = Service.validate_webhook(request, :ignore)
        expected = %URLValidationResponse{code: 0, messages: []}
        assert response == expected
        assert called WebhookSender.validate_webhook(url, "", "")
      end
    end
  end

  defp rule1 do
    %Rule{action: {:SlackAlert, %Notifications.SlackAlert{url: "http://foo.bar:20/baz"}},
          event: 0,
          name: "rule1", id: "rule1_ID"}
  end

  defp rule2_slack do
    %Rule{action: {:SlackAlert, %Notifications.SlackAlert{url: "http://foo.bar:20/baz"}},
          event: 0,
      name: "rule2_slack", id: "rule2_slack_ID"}
  end

  defp rule3_webhook do
    %Rule{action: {:WebhookAlert, %Notifications.WebhookAlert{url: "http://foo.bar:20/baz"}},
          event: 0,
          name: "rule3_webhook", id: "rule3_webhook_ID"}
  end
end

File: components/notifications-service/integration_test/service_test.go
package integration_test

import (
	"context"
	"fmt"
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	deploymentConst "github.com/chef/automate/components/automate-deployment/pkg/constants"
	"github.com/chef/automate/components/notifications-client/api"
	"github.com/chef/automate/lib/grpc/secureconn"
	"github.com/chef/automate/lib/pcmp/passert"
	"github.com/chef/automate/lib/tls/certs"
)

var (
	suite = &Suite{}
	ctx   = context.Background()
)

type Suite struct {
	Client api.NotificationsClient
}

func (s *Suite) GlobalSetup() error {
	// Sometimes in local dev we do partial deployments with just a few services,
	// but the deployment-service should always be there. Using deployment's
	// certs should always work.
	deploymentCerts := certs.TLSConfig{
		CertPath:       deploymentConst.CertPath,
		KeyPath:        deploymentConst.KeyPath,
		RootCACertPath: deploymentConst.RootCertPath,
	}
	certData, err := deploymentCerts.ReadCerts()
	if err != nil {
		return err
	}

	mutTLSFactory := secureconn.NewFactory(*certData)
	grpcEndpoint, err := mutTLSFactory.Dial("notifications-service", "localhost:10125")
	if err != nil {
		return err
	}

	s.Client = api.NewNotificationsClient(grpcEndpoint)

	return s.DeleteEverything()
}

func (s *Suite) GlobalTeardown() error {
	return s.DeleteEverything()
}

func (s *Suite) DeleteEverything() error {
	ruleList, err := s.Client.ListRules(ctx, &api.Empty{})
	if err != nil {
		return err
	}
	for _, r := range ruleList.Rules {
		_, err := s.Client.DeleteRule(ctx, &api.RuleIdentifier{Id: r.Id})
		if err != nil {
			return nil
		}
	}
	return nil
}

// TestMain allow us to run a setup before running our tests and also
// teardown everything after we have finished testing.
//
// (Check out 'suite_test.go')
//
// => Docs: https://golang.org/pkg/testing/#hdr-Main
func TestMain(m *testing.M) {
	// Global Setup hook: Here is where you can initialize anythings you need
	// for your tests to run
	err := suite.GlobalSetup()
	if err != nil {
		fmt.Println("Test suite failed during GlobalSetup")
		fmt.Println(err)
		os.Exit(1)
	}

	// Execute the test suite and record the exit code
	exitCode := m.Run()

	// Teardown hook: It says it all
	err = suite.GlobalTeardown()
	if err != nil {
		fmt.Println("Test suite failed during GlobalTeardown")
		fmt.Println(err)
		if exitCode != 0 {
			os.Exit(2)
		}
	}

	// call with result of m.Run()
	os.Exit(exitCode)
}

/*
	API Endpoints to Test:

	* rpc Notify(Event) returns (Response);
	* rpc ValidateWebhook(URLValidationRequest) returns (URLValidationResponse);
		}
*/

func TestGetVersion(t *testing.T) {
	res, err := suite.Client.Version(ctx, &api.VersionRequest{})
	require.NoError(t, err)
	// In local dev the version string changes to include extra information (in
	// the elixir notifications service). Allow any non-empty string here to be
	// less brittle.
	assert.NotEmpty(t, res.Version)
}

func TestGetRule(t *testing.T) {
	err := suite.DeleteEverything()
	assert.NoError(t, err)

	t.Run("returns not found when requesting a rule that doesn't exist", func(t *testing.T) {
		res, err := suite.Client.GetRule(ctx, &api.RuleIdentifier{Id: "42"})
		require.NoError(t, err)
		assert.Equal(t, api.RuleGetResponse_NOT_FOUND, res.Code)
		assert.Empty(t, res.Rule)
	})

	t.Run("returns the rule when the rule exists", func(t *testing.T) {
		rule := validRule("TestGetRule_success_01")
		createResponse, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		require.NotEmpty(t, createResponse.Id)

		res, err := suite.Client.GetRule(ctx, &api.RuleIdentifier{Id: createResponse.Id})
		require.NoError(t, err)
		assert.Equal(t, api.RuleGetResponse_OK, res.Code)

		rule.Id = createResponse.Id
		passert.Equal(t, rule, res.Rule)
	})
}

func TestListRules(t *testing.T) {
	err := suite.DeleteEverything()
	require.NoError(t, err)

	t.Run("Returns an empty list when no rules exist", func(t *testing.T) {
		ruleList, err := suite.Client.ListRules(ctx, &api.Empty{})
		require.NoError(t, err)
		assert.Empty(t, ruleList.Rules)
	})

	t.Run("Returns the rules when rules exist", func(t *testing.T) {
		expectedRuleNames := []string{"TestListRules_success_01", "TestListRules_success_02"}
		for _, name := range expectedRuleNames {
			_, err := suite.Client.AddRule(ctx, validRule(name))
			require.NoError(t, err)
		}

		ruleList, err := suite.Client.ListRules(ctx, &api.Empty{})
		require.NoError(t, err)
		assert.NotEmpty(t, ruleList.Rules)
		actualRuleNames := []string{}
		for _, r := range ruleList.Rules {
			actualRuleNames = append(actualRuleNames, r.Name)
		}
		assert.ElementsMatch(t, expectedRuleNames, actualRuleNames)
	})
}

func TestAddRuleSuccessCase(t *testing.T) {
	err := suite.DeleteEverything()
	require.NoError(t, err)

	t.Run("valid rule returns a successful response", func(t *testing.T) {
		rule := validRule("TestAddRule_success_01")
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_ADDED, response.Code)
		assert.NotEmpty(t, response.Id)
		passert.Equal(t, rule, response.Rule)
	})

}

func TestAddRuleValidations(t *testing.T) {
	err := suite.DeleteEverything()
	require.NoError(t, err)

	t.Run("the Id field must be empty", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_Id")
		rule.Id = "43"
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "Rule ID may not be included in an add-rule request")
	})

	t.Run("the Name field must not be empty", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_Name")
		rule.Name = ""
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "Rule name must be supplied.")
	})

	t.Run("the Name must be unique", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_unique")
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_ADDED, response.Code)
		assert.NotEmpty(t, response.Id)

		response, err = suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_DUPLICATE_NAME, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A rule with this name already exists")
	})

	t.Run("the Action field must not be empty", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_Action")
		rule.Action = nil
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "Action must be set.")
	})

	t.Run("the URL in a SlackAlert action must not be empty", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_SlackAlertURL")
		rule.GetSlackAlert().Url = ""
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})
	t.Run("the URL in a SlackAlert action must include the protocol/scheme", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_SlackAlertURL")
		rule.GetSlackAlert().Url = "example.com"
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})
	t.Run("the URL in a SlackAlert action must include the hostname", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_SlackAlertURL")
		rule.GetSlackAlert().Url = "https://"
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})

	t.Run("the URL in a WebhookAlert action must not be empty", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_WebhookAlertURL")
		rule.Action = validWebhook()
		rule.GetWebhookAlert().Url = ""
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})
	t.Run("the URL in a WebhookAlert action must include the protocol/scheme", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_WebhookAlertURL")
		rule.Action = validWebhook()
		rule.GetWebhookAlert().Url = "example.com"
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})
	t.Run("the URL in a WebhookAlert action must include the hostname", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_WebhookAlertURL")
		rule.Action = validWebhook()
		rule.GetWebhookAlert().Url = "https://"
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})

	t.Run("the URL in a ServiceNowAlert action must not be empty", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_ServiceNowAlertURL")
		rule.Action = validServiceNowAlert()
		rule.GetServiceNowAlert().Url = ""
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})
	t.Run("the URL in a ServiceNowAlert action must include the protocol/scheme", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_ServiceNowAlertURL")
		rule.Action = validServiceNowAlert()
		rule.GetServiceNowAlert().Url = "example.com"
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})
	t.Run("the URL in a ServiceNowAlert action must include the hostname", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_ServiceNowAlertURL")
		rule.Action = validServiceNowAlert()
		rule.GetServiceNowAlert().Url = "https://"
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "A valid action URL must be supplied")
	})

	t.Run("Multiple error messages are returned when multiple validation rules are not satisfied", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_MuliInvalid")
		rule.Action = nil
		rule.Name = ""
		rule.Id = "42"
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_VALIDATION_ERROR, response.Code)
		assert.Empty(t, response.Id)
		assert.Contains(t, response.Messages, "Action must be set.")
		assert.Contains(t, response.Messages, "Rule name must be supplied.")
		assert.Contains(t, response.Messages, "Rule ID may not be included in an add-rule request")
	})

	// NOTE: this is a test of a validation that doesn't exist
	// NOTE: it's unclear to me if this behavior is correct. It's how the
	// original elixir implementation behaves.
	t.Run("the SecretId in a ServiceNowAlert action can be blank", func(t *testing.T) {
		rule := validRule("TestAddRule_validations_ServiceNowAlert_SecretId")
		rule.Action = validServiceNowAlert()
		rule.GetServiceNowAlert().SecretId = ""
		response, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_ADDED, response.Code)
		assert.NotEmpty(t, response.Id)
		assert.Empty(t, response.Messages)
	})

}

func TestUpdateRuleSuccess(t *testing.T) {
	err := suite.DeleteEverything()
	require.NoError(t, err)

	t.Run("valid rule returns a successful response", func(t *testing.T) {
		rule := validRule("TestAddRule_success_01")
		createResponse, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_ADDED, createResponse.Code)

		rule.Id = createResponse.Id
		rule.GetSlackAlert().Url = "https://newurl.example"
		response, err := suite.Client.UpdateRule(ctx, rule)
		require.NoError(t, err)
		assert.Empty(t, response.Messages)
		assert.Equal(t, api.RuleUpdateResponse_OK, response.Code)
	})

	t.Run("update can modify the rule name", func(t *testing.T) {
		rule := validRule("TestAddRule_success_02")
		createResponse, err := suite.Client.AddRule(ctx, rule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_ADDED, createResponse.Code)

		rule.Id = createResponse.Id
		rule.Name = "TestAddRule_success_02_updated_value"
		response, err := suite.Client.UpdateRule(ctx, rule)
		require.NoError(t, err)
		assert.Empty(t, response.Messages)
		assert.Equal(t, api.RuleUpdateResponse_OK, response.Code)
	})

}

func TestUpdateRuleValidation(t *testing.T) {
	err := suite.DeleteEverything()
	require.NoError(t, err)

	ruleToCreate := validRule("TestUpdateRule_validations")
	created, err := suite.Client.AddRule(ctx, ruleToCreate)
	require.NoError(t, err)
	createdRule := created.Rule
	createdRule.Id = created.Id

	t.Run("the Id field must not be empty", func(t *testing.T) {
		updatedRule := copyRule(createdRule)
		updatedRule.Id = ""
		response, err := suite.Client.UpdateRule(ctx, updatedRule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleUpdateResponse_VALIDATION_ERROR, response.Code)
		assert.Contains(t, response.Messages, "Rule ID must be included from the rule being modified")
	})

	t.Run("the Name field must not be empty", func(t *testing.T) {
		updatedRule := copyRule(createdRule)
		updatedRule.Name = ""
		response, err := suite.Client.UpdateRule(ctx, updatedRule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleUpdateResponse_VALIDATION_ERROR, response.Code)
		assert.Contains(t, response.Messages, "Rule name must be supplied.")
	})

	t.Run("the Name must be unique", func(t *testing.T) {
		rule2 := validRule("TestAddRule_validations_unique")
		res, err := suite.Client.AddRule(ctx, rule2)
		require.NoError(t, err)
		assert.Equal(t, api.RuleAddResponse_ADDED, res.Code)
		assert.NotEmpty(t, res.Id)

		updatedRule := copyRule(createdRule)
		updatedRule.Name = "TestAddRule_validations_unique"
		response, err := suite.Client.UpdateRule(ctx, updatedRule)
		require.NoError(t, err)
		// TODO: the elixir implementation does not check for this case so it
		// returns an internal error when the db request fails on the key constraint.
		// This should be changed so the server detects the duplicate
		assert.NotEqual(t, api.RuleUpdateResponse_OK, response.Code)
		//assert.Equal(t, api.RuleUpdateResponse_DUPLICATE_NAME, response.Code)
		//assert.Contains(t, response.Messages, "A rule with this name already exists")
	})

	t.Run("the Action field must not be empty", func(t *testing.T) {
		updatedRule := copyRule(createdRule)
		updatedRule.Action = nil
		response, err := suite.Client.UpdateRule(ctx, updatedRule)
		require.NoError(t, err)
		assert.Equal(t, api.RuleUpdateResponse_VALIDATION_ERROR, response.Code)
		assert.Contains(t, response.Messages, "Action must be set.")
	})

	// NOTE: UpdateRule is expected to apply the same validations to the actions
	// as AddRule does. We don't explicitly test those because it's repetitive
}

func TestDeleteRule(t *testing.T) {
	err := suite.DeleteEverything()
	require.NoError(t, err)

	ruleToCreate := validRule("TestUpdateRule_validations")
	created, err := suite.Client.AddRule(ctx, ruleToCreate)
	require.NoError(t, err)

	t.Run("deleting an existing rule succeeds", func(t *testing.T) {
		response, err := suite.Client.DeleteRule(ctx, &api.RuleIdentifier{Id: created.Id})
		require.NoError(t, err)
		assert.Equal(t, api.RuleDeleteResponse_DELETED, response.Code)
	})

	t.Run("deleting a rule that doesn't exist returns an error", func(t *testing.T) {
		response, err := suite.Client.DeleteRule(ctx, &api.RuleIdentifier{Id: "5"})
		require.NoError(t, err)
		assert.Equal(t, api.RuleDeleteResponse_NOT_FOUND, response.Code)
	})
}

func copyRule(r *api.Rule) *api.Rule {
	// NOTE: shallow copy. This will work as long as you only modify r2 by
	// assigning a new value for a field.
	r2 := *r
	return &r2
}

func validRule(name string) *api.Rule {
	return &api.Rule{
		Event: api.Rule_CCRFailure,
		Name:  name,
		Action: &api.Rule_SlackAlert{
			SlackAlert: &api.SlackAlert{
				Url: "https://slack.example",
			},
		},
	}
}

func validWebhook() *api.Rule_WebhookAlert {
	return &api.Rule_WebhookAlert{
		WebhookAlert: &api.WebhookAlert{Url: "https://webhook.example"},
	}
}

func validServiceNowAlert() *api.Rule_ServiceNowAlert {
	return &api.Rule_ServiceNowAlert{
		ServiceNowAlert: &api.ServiceNowAlert{
			Url:                  "https://servicenow.example",
			SecretId:             "supersecret",
			CriticalControlsOnly: false,
		},
	}
}
